templates/model/spec.go

```diff
@@ -0,0 +1,285 @@
+package model
+
+// Notes for maintainers explaining the confusing stuff in this file:
+//
+// Q. Why do we need to override UnmarshalYAML()?
+//    A. We want to save the location within the YAML file of each object, so
+//       that we can return helpful error messages that point to the problem.
+//    A. We want to reject unrecognized fields. Due to a known issue in yaml.v3
+//       (https://github.com/go-yaml/yaml/issues/460), this feature doesn't
+//       work in some situations. So we have to implement it ourselves.
+//    A. In the case of the Step struct, we want to do polymorphic decoding
+//       based on the value of the "action" field.
+//
+// Q. What's up with the weird "shadow" unmarshalling pattern (ctrl-f "shadow")
+// A. We often want to "unmarshal all the fields of my struct, but also run
+//    our own logic before and after unmarshaling. To do this, we have to create
+//    a separate type that *doesn't* implement yaml.Unmarshaler, and unmarshal
+//    into that first. If we're inside UnmarshalYAML() and we try to unmarshal
+//    into the receiver, there will be infinite recursion into UnmarshalYAML().
+//    See e.g. https://github.com/go-yaml/yaml/issues/107#issuecomment-524681153.
+//
+// Q. Why is validation done as a separate pass instead of in UnmarshalYAML()?
+// A. Because there's a very specific edge case that we need to avoid.
+//    UnmarshalYAML() is only called for YAML objects that have at least one
+//    field that's specified in the input YAML. This can happen if an object
+//    relies on default values or has no parameters. But we still want to
+//    validate every object. So we need to run validation separately from
+//    unmarshaling.
+
+import (
+	"errors"
+	"fmt"
+	"io"
+
+	"gopkg.in/yaml.v3"
+)
+
+// NewDecoder returns a yaml Decoder with the desired options.
+func NewDecoder(r io.Reader) *yaml.Decoder {
+	dec := yaml.NewDecoder(r)
+	dec.KnownFields(true) // Fail if any unexpected fields are seen. Often doesn't work: https://github.com/go-yaml/yaml/issues/460
+	return dec
+}
+
+// Spec represents a parsed spec.yaml file describing a template.
+type Spec struct {
+	// Pos is the YAML file location where this object started.
+	Pos ConfigPos `yaml:"-"`
+
+	APIVersion String `yaml:"apiVersion"`
+	Kind       String `yaml:"kind"`
+
+	Desc   String   `yaml:"desc"`
+	Inputs []*Input `yaml:"inputs"`
+	Steps  []*Step  `yaml:"steps"`
+}
+
+// UnmarshalYAML implements yaml.Unmarshaler.
+func (s *Spec) UnmarshalYAML(n *yaml.Node) error {
+	if err := extraFields(n, []string{"apiVersion", "kind", "desc", "inputs", "steps"}); err != nil {
+		return err
+	}
+
+	type shadowType Spec
+	shadow := &shadowType{} // unmarshal into a type that doesn't have UnmarshalYAML
+	if err := n.Decode(shadow); err != nil {
+		return err
+	}
+
+	*s = Spec(*shadow)
+	s.Pos = yamlPos(n)
+	return nil
+}
+
+// Validate implements Validator.
+func (s *Spec) Validate() error {
+	return errors.Join(
```

Comment:
I really love how `errors.Join` ignores nil errors - it makes these kinds of things so much nicer.

Reply:
Yeah, it's like an "error monad" or something. Kinda DSL-ish.